bittide/src/Bittide/Counter.hs

```diff
@@ -0,0 +1,134 @@
+-- SPDX-FileCopyrightText: 2022 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+
+module Bittide.Counter
+  ( Active
+  , domainDiffCounter
+  ) where
+
+import Clash.Explicit.Prelude
+
+import Clash.Cores.Xilinx.Xpm (xpmCdcGray)
+import Clash.Sized.Extra (unsignedToSigned, glueUnsigneds)
+
+-- | State of 'domainDiffCounter'
+data DdcState
+  -- | In reset, or waiting for the incoming counter to change
+  = DdcInReset
+  -- | Counting and comparing with incoming domain
+  | DdcRunning (Unsigned 64)
+  deriving (Generic, NFDataX)
+
+-- | Indicates whether 'domainDiffCounter' is actively counting or still in reset.
+type Active = Bool
+
+-- | Determine speed differences between two domains. If the source domain is
+-- faster than the destination domain, the result will become smaller. Vice versa,
+-- if the source domain is slower than the destination domain, the result will
+-- become larger. This is analogous to what would happen to a FIFO's data count
+-- when continuously written to by the source domain and read from by the
+-- destination domain. To ease integration in control algorithms, this component
+-- makes sure it starts counting at zero. It also waits for the incoming counter
+-- to become active, i.e. non-zero, before starting to count itself.
+--
+-- If both domains support initial values, 'domainDiffCounter' does not need to
+-- be reset.
+--
+-- To reset this component, the reset should be asserted for at least one cycle in
+-- the source domain _plus_ four cycles in the destination domain. The reset in the
+-- destination domain should be deasserted at the same time or *after* the one in
+-- the source domain for glitchless operation.
+--
+-- __N.B.__:
+--   This function will only work properly if the given domains are pretty close
+--   to another for a number of reasons:
+--
+--     1. It uses an 8-bit Gray counter internally for CDC
+--     2. It uses one 64-bit counter in each domain
+--     2. Its output is constrained to @Signed 16@
+--
+domainDiffCounter ::
+  forall src dst .
+  ( KnownDomain src
+  , KnownDomain dst
+  ) =>
+  Clock src -> Reset src ->
+  Clock dst -> Reset dst ->
+  -- | Counter and boolean indicating whether the component is currently active
+  Signal dst (Signed 16, Active)
+domainDiffCounter clkSrc rstSrc clkDst rstDst =
+  mealy clkDst rstDst enableGen go DdcInReset counter
+ where
+  -- 64 bits is enough for approximately 3 millenia @ 200 MHz
+  counter = synchronizedSuccCounter @64 clkSrc rstSrc clkDst rstDst
+
+  go :: DdcState -> Unsigned 64 -> (DdcState, (Signed 16, Bool))
+  go DdcInReset c1
+    | c1 == 0   = (DdcInReset,          (0, False))
+    | otherwise = (DdcRunning (c1 + 1), (0, True))
+  go (DdcRunning c0) c1 = (DdcRunning (c0 + 1), (c0 `subAndTruncate` c1, True))
+
+  subAndTruncate :: Unsigned 64 -> Unsigned 64 -> Signed 16
+  subAndTruncate c0 c1 = truncateB (unsignedToSigned c0 - unsignedToSigned c1)
```

Comment:
You're right, I didn't do my back-of-a-napkin calculation right. The goal of this would to basically never overflow in a Bittide system. Given our clock setup we know that the maximum divergence is +-100 ppm. On a 200 MHz clock this amounts to 40k ticks difference at most, every second. A 16 bit number would therefore quickly overflow (~3 seconds).

I think it's therefore best to up this to a `Signed 32`. This would allow a system to run a maximum divergence rate for 2.5 days. This should be more than enough given that neither the rate nor that duration is realistic - clock control should have kicked in way before that.

Reply:
We decided that this is enough for now.